### ISSUE TITLE
Confusing on_val_start and on_val_end callbacks

### ISSUE THREAD

**Issue by wszalapski** [ultralytics/ultralytics#5767]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train, Val

### Bug

The `on_val_start` callback is executed in the `BaseValidator.__call__` only if `self.training` is `False`. On the other hand, the `on_val_end` callback is executed on every `__call__`. Thus, in the provided example, you can see "Validation started" printed only once while "Validation ended" is printed three times (two epochs + final validation). I believe this is counterintuitive not only for me.

Please note that the validation callbacks are added via a workaround. This is related to #3995 which still seem to be not solved and is closed.

### Environment

Ultralytics YOLOv8.0.199 🚀 Python-3.11.5 torch-2.0.0+cu117 CUDA:0 (NVIDIA GeForce RTX 2070 with Max-Q Design, 7974MiB)
Setup complete ✅ (12 CPUs, 31.1 GB RAM, 819.3/914.7 GB disk)

### Minimal Reproducible Example

```python
from ultralytics import YOLO


def val_start_callback(trainer):
    print("Validation started")


def val_end_callback(trainer):
    print("Validation ended")


def add_val_callbacks(trainer):
    trainer.validator.add_callback("on_val_start", val_start_callback)
    trainer.validator.add_callback("on_val_end", val_end_callback)


def train():
    model = YOLO(task="detect")
    model.add_callback("on_train_start", add_val_callbacks)
    model.train(epochs=2, data="coco128.yaml", val=True)


train()
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@wszalapski thank you for raising this issue. The `on_val_start` and `on_val_end` callbacks are indeed designed to be called at different stages of the model training and validation process. The `on_val_start` callback is meant to be executed at the start of each validation, while the `on_val_end` callback is aimed to execute when each validation ends. 

As you have correctly observed, the `on_val_start` is only triggered when `self.training` is `False`; this is because we want to ensure the model is not in training mode before starting validation. As for the `on_val_end` callback, it is designed to execute at the end of each validation iteration, hence being called every time `__call__` is invoked.

However, I understand that the frequency of "Validation ended" getting printed seems counterintuitive, but this behaviour is in line with how these callbacks are designed to function. It seems the report reflects more of an enhancement request than a bug. We can definitely take this into consideration and discuss aligning these callbacks to reflect a more balanced view of the validation process. 

As for the issue you mentioned regarding validation callbacks, #3995, thanks for bringing this to our attention. If you feel that the current implementation is still not providing the functionality you expected, and it's impacting your projects, feel free to open a new issue detailing your exact use-case and requirements.

Again, thanks for reaching out, and your contribution to the YOLOv8 development is appreciated!

**Comment by glenn-jocher** [comment]
@wszalapski yes I'm able to reproduce the same result with your code. Val start and Val end are not equally distributed among the training validation and final validation:

<img width="1508" alt="Screenshot 2023-10-18 at 13 42 15" src="https://github.com/ultralytics/ultralytics/assets/26833433/0faf3343-4c87-4415-9b62-8f3175b1e6fd">

**Comment by wszalapski** [comment]
@glenn-jocher Thank you for the feedback.

>As you have correctly observed, the on_val_start is only triggered when self.training is False; this is because we want to ensure the model is not in training mode before starting validation.

From what I can see, if `self.training` in `BaseValidator` is set to `True`, the validation is still performed. Only the model and data preparation are different. From your latter comment I expect you have already confirmed it happens.

If you would like to keep the current trigger points of the `on_val_start` and `on_val_end` callbacks, it might be worth to denote the difference in the documentation (https://docs.ultralytics.com/usage/callbacks/#validator-callbacks) to help avoid confusion. For a long-term solution, I believe it is good to make them aligned.

**Comment by glenn-jocher** [comment]
@wszalapski I've created a fix for this in PR #5790. This moves the on_val_start callback to the same line for both training-validation and standalone-validation.

**Comment by glenn-jocher** [comment]
@wszalapski You're welcome! I'm glad I could help. If you have any further questions or come across any other issues, don't hesitate to reach out. Your feedback is greatly appreciated and helps us improve YOLOv8. Happy coding!
